File: apps/api/app/recurrences/adapters/outbound/persistence/repository.py
from __future__ import annotations

from datetime import date
from typing import cast

from sqlmodel import Session, select

from app.recurrences.adapters.outbound.persistence.models import RecurrenceModel, utcnow
from app.recurrences.domain.entities import (
    DraftRecurrence,
    Frequency,
    Recurrence,
    RecurrenceKind,
)
from app.recurrences.domain.errors import RecurrenceNotFoundError
from app.shared.domain.money import Money


class SqlRecurrenceRepository:
    """Satisfies the RecurrenceRepository port (structural typing)."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def add(self, draft: DraftRecurrence) -> Recurrence:
        model = RecurrenceModel(
            kind=draft.kind,
            amount_cents=draft.money.amount_cents,
            currency=draft.money.currency,
            category_id=draft.category_id,
            frequency=draft.frequency,
            day_of_month=draft.day_of_month,
            second_day_of_month=draft.second_day_of_month,
            start_date=draft.start_date,
            end_date=draft.end_date,
            note=draft.note,
            is_active=draft.is_active,
        )
        self._session.add(model)
        self._session.commit()
        self._session.refresh(model)
        return _to_entity(model)

    def get(self, recurrence_id: int) -> Recurrence | None:
        model = self._session.get(RecurrenceModel, recurrence_id)
        if model is None or model.deleted_at is not None:
            return None
        return _to_entity(model)

    def list_all(self) -> list[Recurrence]:
        statement = select(RecurrenceModel).where(RecurrenceModel.deleted_at.is_(None))  # type: ignore[union-attr]
        return [_to_entity(row) for row in self._session.exec(statement).all()]

    def list_active(self) -> list[Recurrence]:
        statement = select(RecurrenceModel).where(
            RecurrenceModel.deleted_at.is_(None),  # type: ignore[union-attr]
            RecurrenceModel.is_active.is_(True),  # type: ignore[union-attr]
        )
        return [_to_entity(row) for row in self._session.exec(statement).all()]

    def update(self, recurrence: Recurrence) -> Recurrence:
        model = self._session.get(RecurrenceModel, recurrence.id)
        if model is None or model.deleted_at is not None:
            raise RecurrenceNotFoundError(recurrence.id)
        model.amount_cents = recurrence.money.amount_cents
        model.currency = recurrence.money.currency
        model.category_id = recurrence.category_id
        model.frequency = recurrence.frequency
        model.day_of_month = recurrence.day_of_month
        model.second_day_of_month = recurrence.second_day_of_month
        model.start_date = recurrence.start_date
        model.end_date = recurrence.end_date
        model.note = recurrence.note
        model.is_active = recurrence.is_active
        model.updated_at = utcnow()
        self._session.add(model)
        self._session.commit()
        self._session.refresh(model)
        return _to_entity(model)

    def soft_delete(self, recurrence_id: int) -> bool:
        model = self._session.get(RecurrenceModel, recurrence_id)
        if model is None or model.deleted_at is not None:
            return False
        model.deleted_at = utcnow()
        self._session.add(model)
        self._session.commit()
        return True

    def mark_generated(self, recurrence_id: int, last_generated_on: date) -> None:
        model = self._session.get(RecurrenceModel, recurrence_id)
        if model is None:
            return
        model.last_generated_on = last_generated_on
        model.updated_at = utcnow()
        self._session.add(model)
        self._session.commit()


def _to_entity(model: RecurrenceModel) -> Recurrence:
    assert model.id is not None
    return Recurrence(
        id=model.id,
        kind=cast(RecurrenceKind, model.kind),
        money=Money(model.amount_cents, model.currency),
        category_id=model.category_id,
        frequency=cast(Frequency, model.frequency),
        day_of_month=model.day_of_month,
        second_day_of_month=model.second_day_of_month,
        start_date=model.start_date,
        end_date=model.end_date,
        note=model.note,
        is_active=model.is_active,
        last_generated_on=model.last_generated_on,
    )

File: apps/api/internal/recurrences/adapters/persistence/repository.go
package persistence

import (
	"context"
	"time"

	platformdb "caudal-api/internal/platform/persistence"
	"caudal-api/internal/recurrences/domain"
	shared "caudal-api/internal/shared/domain"
	"gorm.io/gorm"
)

type Repository struct {
	db *gorm.DB
}

func NewRepository(db *gorm.DB) Repository {
	return Repository{db: db}
}

func (repo Repository) Create(ctx context.Context, item domain.Recurrence) (domain.Recurrence, error) {
	model := fromDomain(item)
	if err := repo.db.WithContext(ctx).Create(&model).Error; err != nil {
		return domain.Recurrence{}, err
	}
	return toDomain(model)
}

func (repo Repository) Get(ctx context.Context, id int64) (domain.Recurrence, error) {
	var model RecurrenceModel
	err := repo.active(ctx).First(&model, "id = ?", id).Error
	return mapOne(model, err)
}

func (repo Repository) List(ctx context.Context, kind *domain.Kind) ([]domain.Recurrence, error) {
	var models []RecurrenceModel
	query := repo.active(ctx).Order("id DESC")
	if kind != nil {
		query = query.Where("kind = ?", string(*kind))
	}
	if err := query.Find(&models).Error; err != nil {
		return nil, err
	}
	return toDomains(models)
}

func (repo Repository) ListActive(ctx context.Context) ([]domain.Recurrence, error) {
	var models []RecurrenceModel
	err := repo.active(ctx).Where("is_active = ?", true).Order("id").Find(&models).Error
	if err != nil {
		return nil, err
	}
	return toDomains(models)
}

func (repo Repository) Update(ctx context.Context, item domain.Recurrence) (domain.Recurrence, error) {
	err := repo.db.WithContext(ctx).Model(&RecurrenceModel{}).
		Where("id = ? AND deleted_at IS NULL", item.ID).
		Updates(map[string]any{
			"amount_cents":        item.Money.AmountCents,
			"currency":            item.Money.Currency,
			"category_id":         item.CategoryID,
			"frequency":           string(item.Frequency),
			"day_of_month":        item.DayOfMonth,
			"second_day_of_month": item.SecondDayOfMonth,
			"start_date":          item.StartDate,
			"end_date":            item.EndDate,
			"note":                item.Note,
			"is_active":           item.IsActive,
			"updated_at":          time.Now().UTC(),
		}).Error
	if err != nil {
		return domain.Recurrence{}, err
	}
	return repo.Get(ctx, item.ID)
}

func (repo Repository) Delete(ctx context.Context, id int64) error {
	now := time.Now().UTC()
	result := repo.db.WithContext(ctx).Model(&RecurrenceModel{}).
		Where("id = ? AND deleted_at IS NULL", id).Update("deleted_at", now)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return domain.ErrNotFound
	}
	return nil
}

func (repo Repository) MarkGenerated(ctx context.Context, id int64, date time.Time) error {
	return repo.db.WithContext(ctx).Model(&RecurrenceModel{}).
		Where("id = ? AND deleted_at IS NULL", id).
		Updates(map[string]any{"last_generated_on": date, "updated_at": time.Now().UTC()}).Error
}

func (repo Repository) active(ctx context.Context) *gorm.DB {
	return repo.db.WithContext(ctx).Where("deleted_at IS NULL")
}

func mapOne(model RecurrenceModel, err error) (domain.Recurrence, error) {
	if platformdb.IsNotFound(err) {
		return domain.Recurrence{}, domain.ErrNotFound
	}
	if err != nil {
		return domain.Recurrence{}, err
	}
	return toDomain(model)
}

func fromDomain(item domain.Recurrence) RecurrenceModel {
	return RecurrenceModel{
		ID: item.ID, Kind: string(item.Kind), AmountCents: item.Money.AmountCents,
		Currency: item.Money.Currency, CategoryID: item.CategoryID,
		Frequency: string(item.Frequency), DayOfMonth: item.DayOfMonth,
		SecondDayOfMonth: item.SecondDayOfMonth, StartDate: item.StartDate,
		EndDate: item.EndDate, Note: item.Note, IsActive: item.IsActive,
		LastGeneratedOn: item.LastGeneratedOn, UpdatedAt: time.Now().UTC(),
	}
}

func toDomain(model RecurrenceModel) (domain.Recurrence, error) {
	money, err := shared.NewMoney(model.AmountCents, model.Currency)
	if err != nil {
		return domain.Recurrence{}, err
	}
	return domain.NewRecurrence(domain.Recurrence{
		ID: model.ID, Kind: domain.Kind(model.Kind), Money: money,
		CategoryID: model.CategoryID, Frequency: domain.Frequency(model.Frequency),
		DayOfMonth: model.DayOfMonth, SecondDayOfMonth: model.SecondDayOfMonth,
		StartDate: model.StartDate, EndDate: model.EndDate, Note: model.Note,
		IsActive: model.IsActive, LastGeneratedOn: model.LastGeneratedOn,
	})
}

func toDomains(models []RecurrenceModel) ([]domain.Recurrence, error) {
	items := make([]domain.Recurrence, 0, len(models))
	for _, model := range models {
		item, err := toDomain(model)
		if err != nil {
			return nil, err
		}
		items = append(items, item)
	}
	return items, nil
}
